runtime/lua/vim/lsp/protocol.lua

```diff
@@ -0,0 +1,437 @@
+end
+
+protocol.Position = function(args)
+  args = check_table(args)
+
+  return {
+    line = protocol.line(args.line),
+    character = protocol.character(args.character),
+  }
+end
+
+protocol.TextDocumentPositionParams = function(args)
+  args = check_table(args)
+
+  return {
+    textDocument = protocol.TextDocumentIdentifier(args.textDocument),
+    position = protocol.Position(args.position),
+  }
+end
+
+protocol.ReferenceContext = function(args)
+  args = check_table(args)
+
+  return {
+    includeDeclaration = args.includeDeclaration or true,
+  }
+end
+
+protocol.CompletionContext = function(args)
+  args = check_table(args)
+
+  if args.triggerKind == nil and args.triggerCharacter == nil then
+    return nil
+  end
+
+  return {
+    triggerKind = args.triggerKind or nil,
+    triggerCharacter = args.triggerCharacter or nil,
+  }
+end
+
+-- TODO: Not implement workspace features now.
+protocol.WorkspaceClientCapabilities = {}
+-- {
+--   applyEdit = boolean,
+--   workspaceEdit = {
+--     documentChanges = boolean,
+--     resourceOperations = ResourceOperationKind[],
+--     failureHandling = FailureHandlingKind,
+--   },
+--   didChangeConfiguration = {
+--     dynamicRegistration = boolean,
+--   },
+--   didChangeWatchedFiles = {
+--     dynamicRegistration = boolean,
+--   },
+--     symbol = {
+--     dynamicRegistration = boolean,
+--     symbolKind = {
+--       valueSet = SymbolKind[],
+--     },
+--   },
+--   executeCommand = {
+--     dynamicRegistration = boolean;
+--   },
+--   workspaceFolders = boolean,
+--   configuration = boolean,
+-- }
+
+protocol.TextDocumentClientCapabilities = {
+  synchronization = {
+    dynamicRegistration = false,
+
+    -- Send textDocument/willSave before saving (BufWritePre)
+    willSave = true,
+
+    -- TODO(tjdevries): Implement textDocument/willSaveWaitUntil
+    willSaveWaitUntil = false,
+
+    -- Send textDocument/didSave after saving (BufWritePost)
+    didSave = true,
+  },
+  completion = {
+    dynamicRegistration = false,
+    completionItem = {
+
+      -- TODO(tjdevries): Is it possible to implement this in plain lua?
+      snippetSupport = false,
+      commitCharactersSupport = false,
+      documentationFormat = {'plaintext'},
+    },
+    completionItemKind = {
+      valueSet = vim.tbl_keys(protocol.CompletionItemKind),
+    },
+
+    -- TODO(tjdevries): Implement this
+    contextSupport = false,
+  },
+  hover = {
+    dynamicRegistration = false,
+
+    -- Currently only support plaintext
+    --    In the future, if we have floating windows or display in a preview window,
+    --    we could say markdown
+    contentFormat = {'plaintext'},
+  },
+  signatureHelp = {
+    dynamicRegistration = false,
+    signatureInformation = {
+      documentationFormat = {'plaintext'}
+    },
+  },
+  references = {
+    dynamicRegistration = false,
+  },
+  documentHighlight = {
+    dynamicRegistration = false
+  },
+}
+
+protocol.ClientCapabilities = {
```

Comment:
There seems to be some code included for proper utf-8/utf-16 support (which is nice). This should include `offsetEncoding = {'utf-8', 'utf-16'}` then.

Reply:
But there is not official offsetEncoding spec.
So I think that we should provide the offset encoding option to the `add_server_config` args until it is formulated as a specification.

Reply:
Though if you just add `offset_encoding="utf-8"` it will not work with a server that uses offsetEncoding like clangd. Perhaps `offsetEncoding`  should be used when the option is set? 

Though it would still be nice to use UTF-8 automatically for a server like clangd. Even if it is not in the spec, it is a conservative extension in the sense that it will not break any server that fellows the spec strictly (i e if server doesn't have offsetEncoding, then we follow the spec behavior).

Reply:
The offsetEncoding property is custom client capabilities for clangd.
So I think user should set that like this.

```
call lsp#add_server_config({
	\ "server_name": "clangd",
	\ "filetype": ["c", "cpp"],
	\ "cmd": "clangd -background-index"
	\ 'server_config': {
	\   'offsetEncoding': ['utf-8', 'utf-16'],
	\ }})
```

And if response has a 'offsetEncoding' property, we overrides client.offset_encoding value.

Reply:
makes sense, though `server_config` subdict should probably be called `capabilities` or something thereabout, the entire call is the config.

Reply:
That's nice point. I'll rename.